plaso/engine/artifacts_trie.py

```diff
@@ -0,0 +1,297 @@
+    self.artifacts_names = []
+    self.children = {}
+    self.is_root = is_root
+    self.path_separator = path_separator
+
+
+class ArtifactsTrie(object):
+  """Trie structure for storing artifact paths.
+
+  Attributes:
+    artifacts_paths (dict[str, list[str]]): Artifact paths for glob expansion,
+        keyed by artifact name.
+    root (TrieNode): Root node of the Trie.
+  """
+
+  def __init__(self):
+    """Initializes an artifact trie object."""
+    super(ArtifactsTrie, self).__init__()
+    self.artifacts_paths = {}
+    self.root = TrieNode(is_root=True)
+
+  def AddPath(self, artifact_name, path, path_separator):
+    """Adds a path from an artifact definition to the Trie.
+
+    Args:
+      artifact_name (str): name of the artifact.
+      path (str): path from the artifact definition.
+      path_separator (str): path separator.
+    """
+    logger.debug(f'Adding path: "{path:s}" to artifact: "{artifact_name:s}"')
+    self.artifacts_paths.setdefault(artifact_name, []).append(path)
+
+    # Start at the root
+    node = self.root
+    # Add a path separator node if this is a new separator
+    if path_separator not in node.children:
+      node.children[path_separator] = TrieNode(path_separator=path_separator)
+    node = node.children[path_separator]
+    # Handle the case when the input path is equal to the path_separator
+    if path == path_separator:
+      node.artifacts_names.append(artifact_name)
+      return
+
+    path_segments = path.strip(path_separator).split(path_separator)
+    for segment in path_segments:
+      # Store the path_separator for each node.
+      if not hasattr(node, 'path_separator'):
+        node.path_separator = path_separator
+
+      if segment not in node.children:
+        node.children[segment] = TrieNode(path_separator)
+      node = node.children[segment]
+    node.artifacts_names.append(artifact_name)
+
+  def GetMatchingArtifacts(self, path, path_separator):
+    """Retrieves the artifact names that match the given path.
+
+    Args:
+        path (str): path to match against the Trie.
+        path_separator (str): path separator.
+
+    Returns:
+      list[str]: artifact names that match the path.
+    """
+    # Start at the root's child that matches the path_separator
+    if path_separator not in self.root.children:
+      return []
+
+    sub_root_node = self.root.children[path_separator]
+    # Handle the case when the input path is equal to the path_separator
+    if path == path_separator:
+      matching_artifacts = set()
+      if sub_root_node.artifacts_names:
+        matching_artifacts.update(sub_root_node.artifacts_names)
+      return list(matching_artifacts)
+    path_segments = path.strip(path_separator).split(path_separator)
+    matching_artifacts = set()
+    # Update self.artifacts_paths before starting the search.
+    self.artifacts_paths = self._GetArtifactsPaths(sub_root_node)
+
+    def _search_trie(node, current_path, segments):
```

Comment:
Is it possible to make this a classmethod ?

Reply:
Done